gateway/conversation_marking.go

```diff
@@ -0,0 +1,84 @@
+package gateway
+
+import (
+	"log"
+	"sync"
+	"time"
+
+	"github.com/nlopes/slack"
+)
+
+// ConversationMarker handles coordination and de-duplication of marking conversations as read
+type ConversationMarker struct {
+	rtmIDToMarkFuncMap          map[int]func(string)
+	conversationToReadCursorMap map[string]string
+
+	sync.Mutex
+}
+
+// NewConversationMarker creates a new conversation marker
+func NewConversationMarker() *ConversationMarker {
+	return &ConversationMarker{
+		rtmIDToMarkFuncMap:          map[int]func(string){},
+		conversationToReadCursorMap: map[string]string{},
+	}
+}
+
+func (cm *ConversationMarker) markConversation(conversationID, timestamp string, mark func(string, string) error) {
+	cm.conversationToReadCursorMap[conversationID] = timestamp
+
+	go func() {
+		time.Sleep(5 * time.Second)
```

Comment:
Why don't we check if `timestamp > cm.conversationToReadCursorMap[conversationId]`? If you're worried about this being on the synchronous path, we can spawn a goroutine when calling `markConversation`

Reply:
not sure what you're suggesting here. despite the name, `timestamp` apparently isn't actually a timestamp:

![The message thread data model](https://pshuu.moe/2pk/g8gwVY.png)

So while they can be ordered, apparently you don't want to order them...